src/main/java/carpet/script/CarpetExpression.java

```diff
@@ -2096,21 +2096,29 @@ private void API_IteratingOverAreasOfBlocks()
             return (c_, t_) -> new NumericValue(finalSCount);
         });
 
-        this.expr.addLazyFunction("volume", 7, (c, t, lv) ->
-        {
-            int xi = (int) NumericValue.asNumber(lv.get(0).evalValue(c)).getLong();
-            int yi = (int) NumericValue.asNumber(lv.get(1).evalValue(c)).getLong();
-            int zi = (int) NumericValue.asNumber(lv.get(2).evalValue(c)).getLong();
-            int xj = (int) NumericValue.asNumber(lv.get(3).evalValue(c)).getLong();
-            int yj = (int) NumericValue.asNumber(lv.get(4).evalValue(c)).getLong();
-            int zj = (int) NumericValue.asNumber(lv.get(5).evalValue(c)).getLong();
-            int minx = min(xi, xj);
-            int miny = min(yi, yj);
-            int minz = min(zi, zj);
-            int maxx = max(xi, xj);
-            int maxy = max(yi, yj);
-            int maxz = max(zi, zj);
-            LazyValue expr = lv.get(6);
+        this.expr.addLazyFunction("volume", -1, (c, t, lv) ->
+        {
+            CarpetContext cc = (CarpetContext)c;
+
+            if(lv.size()!=3 & lv.size()!=5 & lv.size()!=7) throw new InternalExpressionException("'volume' takes 3, 5 or 7 arguments.");
+
+            BlockArgument pos1Locator = BlockArgument.findIn(cc, lv, 0);
+            BlockPos pos2 = BlockArgument.findIn(cc, lv, pos1Locator.offset).block.getPos();
+            BlockPos pos1 = pos1Locator.block.getPos();
+
+            int x1 = pos1.getX();
+            int y1 = pos1.getY();
+            int z1 = pos1.getZ();
+            int x2 = pos2.getX();
+            int y2 = pos2.getY();
+            int z2 = pos2.getZ();
+            int minx = min(x1, x2);
+            int miny = min(y1, y2);
+            int minz = min(z1, z2);
+            int maxx = max(x1, x2);
+            int maxy = max(y1, y2);
+            int maxz = max(z1, z2);
+            LazyValue expr = lv.get(lv.size()-1);
```

Comment:
likewise, you should use pos2.offset to assert you have a correct argument, not size-1